searx/engines/sqlite.py

```diff
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+# pylint: disable=missing-function-docstring
+
+"""SQLite database (Offline)
+
+"""
+
+import sqlite3
+
+from searx import logger
+
+logger = logger.getChild('SQLite engine')
+
+engine_type = 'offline'
+database = ""
+query_str = ""
+limit = 10
+paging = True
+result_template = 'key-value.html'
+
+class SQLiteDB:
```

Comment:
IMO the function version is more readable:
```python
from contextlib import contextmanager

@contextmanager
def sqlite_cursor(database_url):
	try:
		connect = sqlite3.connect(database_url)
		connect.row_factory = sqlite3.Row
		cursor = self.connect.cursor()
		try:
			yield cursor
		finally:
			cursor.close()
	finally:
		if connect:
			connect.close()
```

Reply:
Hm if you insist we can use a decorator .. I'm more used to use classes.  And I'm not a  fan boy "exceptional programming" .. try blocks are expensive and often miss used I think. Compare, less indentations no try block needed:

```python
class SQLiteCursor:
    def __init__(self, db):
        self.database = db
        self.connect = None
        self.cursor = None
    def __enter__(self):
        self.connect = sqlite3.connect(self.database)
        self.connect.row_factory = sqlite3.Row
        self.cursor = self.connect.cursor()
        return self.cursor
    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_type is None:
            self.connect.commit()
        self.cursor.close()
        self.connect.close()
```

Reply:
Scenario:
* some queries crash (whatever the reason)
* cursor and connect are not closed right away but with GC.

The try...finally blocks are here to make the connection whatever happens.

If the cursor is closed with the database it can be:
```python
@contextmanager
def sqlite_cursor(database_url):
	try:
		connect = sqlite3.connect(database_url)
		connect.row_factory = sqlite3.Row
		yield self.connect.cursor()
	finally:
		if connect:
			connect.close()
```

But everything can be fit the `search` function.

Reply:
> Searches are read-only operations. I don't think a SQL COMMIT is required here.

see my https://github.com/searxng/searxng/pull/80#discussion_r636230676

Reply:
> Scenario:
>    some queries crash (whatever the reason)

May I do not understand all .. AFAIK for this we use the "with" statement.  Can't speak for the decorator solution, but the `__exit__` method of the class is called,  no matter what was the reason for leaving the "with"" block.

> But everything can be fits into the search function.

Hm .. I do not see the benefit or I misunderstood you.

To get things done: do you like to patch on top what ever solution you prefer .. I can reorder the history afterwards.